doc/architecture/decisions/0013-signals-listeners-model-and-api.md

```diff
@@ -0,0 +1,55 @@
+# 13. Signal/Listener model and API
+
+Date: 2018-05-16
+
+## Status
+
+Proposed
+
+## Context
+
+In the go code we used polling from the UI to discover local state changes, be they in the source-chain or in the DHT.  We want to change this to an event based model where UI (or other clients, like bridged apps) can become listeners to signals.  Additionally there are system level signals (like warrants) as well as other app level definable signals that we want consumers of exposed holochain app APIs to be able to access on a push (i.e. listening) basis.
+
+Note that this is only about signals that can be listened to by container/composer context, i.e. by the client of the core_api, NOT by the app nodes themselves.
+
+## Decision
+
+We will extend the API in a way that's roughly equivalent to the [Signal-slot pattern](https://en.wikipedia.org/wiki/Signals_and_slots), i.e. in the following ways:
+
+1. In the DNA you can declare observable signals anywhere you can declare functions.  You can think of this almost identically to declaring a function except that it "goes the other way," i.e. a function def exposes an entry point where an signal def exposes an exit point.  It would look like this:
+
+``` javascript
+          "signal_declarations": [
```

Comment:
The term `signal` is something originates in the Qt world. I would use a more widely used and adopted term like `event`.

Reply:
Well, not just the Qt world.  It's also a big term in information theory and muti-agent systems and complex adaptive systems worlds, which we are (and I would say) want to be connected to.

Reply:
Signals/Slots is a design pattern that was spearheaded by Qt but there are several other implementations as well, like for instance in boost: https://www.boost.org/doc/libs/1_63_0/doc/html/signals.html.

There is a subtle difference between events and signals/slots:
- with events, the focus is basically on the event object which can exist on its own
- signals/slots is a way to do dynamic binding of function calls, i.e. the use of ```emit``` is actually to be seen as a call to another (set of) function(s).

The reason I prefer the signal/slot frame over events is that it emphasizes the notion of a strong object-model: you connect slots to signals where both belong to a given entity (subclasses of QObject in the case of Qt). There is no signal dangling in the event-loop that does not have a clear source ("sender").

Signal/slots imply an underlying event-loop though on which they are implemented, but are a more elaborate kind of thing.

Since we do have a strong representation of the sender of an event (an app instance), we can make advantage of this more specific concept which I think communicates the right contextual ideas.

Reply:
I agree with @Connoropolous's comment in the main thread: get rid of "_declarations"